Issue 2452

PRITI1999 opened the pull request.

PRITI1999 commented:
@Maxi17 The latest commit fixes the issue. I tried to add a better testcase to ensure `length` and `getLength()` behave similarly. Please provide your feedback.

Maxi17 commented:
@PRITI1999 It looks good. Please fix the misc tests by adding documentation to the methods you added in `ValueMethodIdentifier` and `ValueTransfer`, and then open the PR to the main repository.

mernst committed:
Enforce the `@RelevantJavaTypes` annotation (#3787)

Fixes #1908. Fixes #2075.

kelloggm committed:
`@MatchesRegex` annotation for Constant Value Checker

mernst committed:
Assert that test directories exist

mernst committed:
Reduce latency in Azure Pipelines CI job

mernst committed:
Put visualization code in its own package, to improve code organization

mernst committed:
Re-enable require-javadoc

mernst committed:
Disable `require-javadoc` with a file rather than a script variable

mernst committed:
Fix 3 determinism bugs: 2 in ControlFlowGraph.java, 1 in DOTCFGVisualizer.java

mernst committed:
Make Block extend UniqueId

mernst committed:
Fix 24 determinism bugs: 5 in CFGBuilder, 19 in CFGTranslationPhaseOne

mernst committed:
Better way to prevent instantiation (#3801)

mernst committed:
Abstract out getSimpleClassName() and getMethodName() methods

smillst committed:
Add support of boolean literals to the FlowExpression parser. (#3805)

mernst committed:
Don't test unmaintained project

mernst committed:
Add task to run JSpecify tests

mernst committed:
Make LiveVarStore deterministic

dependabot-preview[bot] committed:
Bump com.gorylenko.gradle-git-properties from 2.2.3 to 2.2.4

dependabot-preview[bot] committed:
Bump net.ltgt.errorprone from 1.2.1 to 1.3.0

mernst committed:
Mention SpotBugs as well as FindBugs

mernst committed:
Fix Javadoc

smillst committed:
Correct AnnotatedIntersectionTypes (#3790)

Defaulting rules for the primary annotation of an intersection type:
1. One or more of the bounds have an explicit annotation: the default annotation is the first explicit annotation.  (I tried making the default the glb of the (explicit or defaulted) annotations on the bounds, but I found this confusing in the case of casts.  We can discuss this further.)
2. Zero bounds have an explicit annotation: the default annotation of the location is applied.  (So, either explicit upper bound defaults, or for casts, the annotation of the cast expression.)

Warning:
If any explicit annotation on a bound is not the same as the primary annotation, then a warning is issued.  (I was going to put this in a separate pull request, but it needs to be in this one so that the test cases are easier to understand.)  

Changes to AnnotatedIntersectionType:
1. Adding a primary annotation to an intersection type causes that annotation to be copied to its bounds.  This is similar to what happens when an annotation is added to a type variable.
2. New method `getBounds` returns a list of the bounds of the intersection type.  (This used to be done via `directSuperTypes` method.
3.  The bounds can be any AnnotatedTypeMirror if the intersection arises from capture conversion.

 Fixes #868; Fixes #3349; Fixes #3362; Fixes #3659.

mernst committed:
"fbc" warring suppression string suppresses initialization errors

mernst committed:
Fix typo

mernst committed:
Improve instructions for bug reporting (#3813)

smillst committed:
Handle synthetic arguments to anonymous class constructors; fixes #3776

mernst committed:
Refactor isAccessAllowed to checkAccessAllowed

mernst committed:
Clarify Initialization Checker documentation.  Fixes #3760.

mernst committed:
Improve the manual's introduction (#3815)

mernst committed:
Remove "Fbc" where not needed to distinguish from Rawness (#3816)

mernst committed:
Put "found" and "required" before long method signature in error messages; fixes #3637 (#3817)

mernst committed:
Note about -AatfDoNotCache

mernst committed:
Override both versions of finishValue (#3820)

mernst committed:
Add new method AnnotationUtils.toStringSimple

mernst committed:
Avoid branches of typetools repositories

mernst committed:
Update URL

mernst committed:
Clarify relationship to upstream versions and remove a file; fixes #3819

mernst committed:
Move treeUids from CFGBuilder to TreeUtils

mernst committed:
Fix determinism bugs in Node and DOTCFGVisualizer

mernst committed:
Make shallow clones

mernst committed:
Make unique IDs per-class rather than global

mernst committed:
Make MostlySingleton deterministic (fixes problems with sets of labels)

mernst committed:
Make `escapeDoubleQuotes` methods static

mernst committed:
Make ConstantPropagationStore deterministic; fixes 4 determinism defects

mernst committed:
Trivial implementation exists to avoid warning message

mernst committed:
Typo fix

dependabot-preview[bot] committed:
Bump guava from 29.0-jre to 30.0-jre

mernst committed:
Import Store.FlowRule (#3836)

mernst committed:
Abstract out fields to avoid re-computation

mernst committed:
Fix Javadoc errors

t-rasmud committed:
Makes HasQualifierParameter an inherited type; Fixes #3831 (#3832)

mernst committed:
Tweak comments

mernst committed:
Annotate `isPoly()` as `@Pure`

mernst committed:
Enrich treatment of `@PolyNull`; fixes #3614

mernst committed:
sideEffectsUnrefineAliases feature

mernst committed:
Don't run javadoc on plume-lib projects

smillst committed:
Add changelog.

smillst committed:
Don't link to BaseTypeVisitor#visitAnnotatedType because it cause javadoc to crash.

smillst committed:
Don't link to a non-public member.

smillst committed:
new release 3.7.1

smillst committed:
Udpate version for next release.

mernst committed:
Don't downcase checker name in user-written SuppressWarnings strings

The feature was undocumented and led to inconsistent `@SuppressWarnings` in user code.

mernst committed:
Redundantly give instructions about how to contribute

mernst committed:
Add argument to MostlyNoElementQualifierHierarchy operations (#3846)

smillst committed:
Use relative path when passing filenames to reformat and checkFormat

mernst committed:
Remove undesired \newpage

mernst committed:
Reuse existing initializer rather than re-implementing

mernst committed:
Rename variable

jwaataja committed:
Don't clear type parameters in StubParser

mernst committed:
Abstract out `forwardsArguments` method (#3856)

csgordon committed:
Fix overriding checks; fixes #3287

mernst committed:
Fix logic bug in comparing arguments to parameters

mernst committed:
Print declaration annotations to stub files

mernst committed:
Define classes `SubtypeIsSubsetQualifierHierarchy` and `SubtypeIsSupersetQualifierHierarchy`

mernst committed:
Define glb more precisely for the Nullness Checker (#3854)

smillst committed:
Reset uninferredTypeArgumentScanner; fixes #3851

mernst committed:
Renamings in ElementUtils: getVerboseName => getQualifiedName, getSimpleName => getSimpleSignature (#3860)

mernst committed:
Documentation improvements

mernst committed:
getContainingMethod returns null for lambdas; fixes issue #3850

mernst committed:
In `updateFromReturn`, remove Java expressions that contain local variables

mernst committed:
New method ElementUtils.getSimpleNameOrDescription

mernst committed:
Remove SKIP-REQUIRE-JAVADOC file

mernst committed:
Remove unnecessary instructions

mernst committed:
Fix formatting of type annotations

mernst committed:
Distinguish StubParser tests

kelloggm committed:
make ElementUtils#getQualifiedName unique for method parameters

mernst committed:
Rename a field and avoid recomputation

mernst committed:
How to build with sbt

Thanks to Suminda Sirinath Salpitikorala Dharmasena.  Fixes #3862.

mernst committed:
Documentation tweaks

mernst committed:
Documentation tweaks

mernst committed:
Improve section labels, add subsections

mernst committed:
Move Aliasing Checker chapter later in manual

mernst committed:
Use description environment for qualifiers

mernst committed:
How to enforce accumulation-related properties (#3874)

mernst committed:
Clarify when a checker shorthand is applied (#3880)

mernst committed:
Expand diagnostics for AnnotatedTypeFactory (#3881)

mernst committed:
Remove unused formal parameters from checkPostcondition and checkConditionalPostcondition

mernst committed:
Infer method declaration annotations

mernst committed:
Explain defaulting for exception parameters

mernst committed:
Move ContractsUtils instance from visitor to type factory

mernst committed:
Add documentation and split a conditional

mernst committed:
Define method getDefaultValueAnnotatedType

mernst committed:
Fix cut-and-paste error

mernst committed:
Simplify -AwarnUnneededSuppressions specification and implementation (#3885)

mernst committed:
Fix indentation

mernst committed:
Handle code that is not in a method; fixes #3884 (#3886)

Co-authored-by: Suzanne Millstein <smillst@cs.washington.edu>

mernst committed:
jtreg tests are nonJunitTests

mernst committed:
Remove unneeded code

Co-authored-by: Suzanne Millstein <smillst@cs.washington.edu>

mernst committed:
Add test case for issue #3888

mernst committed:
Update changelog

mernst committed:
Fix LaTeX error

mernst committed:
Update URL

mernst committed:
Make method.guarantee.violated message more concise and readable

mernst committed:
Initialized Fields Checker

mernst committed:
Add another example invocation

mernst committed:
Add documentation

dependabot-preview[bot] committed:
Bump plume-util from 1.1.7 to 1.2.0

mernst committed:
Improve documentation for ignoreNullAssignments (#3894)

mernst committed:
Increase output from -AstubDebug (#3895)

mernst committed:
Clarify when "parameter" excludes the receiver (#3896)

msridhar committed:
Add AssignmentContexts for parameters of constructor calls (#3891)

mernst committed:
Use `removeIf` instead of constructing a new list

mernst committed:
Rename update methods

mernst committed:
Don't distinguish methods and constructors in purity messages

mernst committed:
Make documentation more informative (#3899)

mernst committed:
Move `getAnnotationsField()` adjacent to `getAnnotations()`

mernst committed:
Compute which processors were auto-discovered

smillst committed:
AnnotatedIntersectionTypes can have TypeVariable bounds (#3904)

smillst committed:
Reduce the length of toString on AnnotatedTypeVariables that are captured types (#3905)

smillst committed:
Substitute type variables in underlying TypeMirrors of AnnotatedTypeMirrors (#3906)

mernst committed:
createType produces an AnnotatedTypeMirror with no annotations

mernst committed:
Use lowercase message keys for consistency (#3909)

Fixes #3908.

mernst committed:
Put dependencies in canonical order (#3910)

mernst committed:
Remove workaround for fixed issue https://github.com/javaparser/javaparser/issues/2615 (#3911)

smillst committed:
Fix format string. (#3912)

smillst committed:
checkCompilerMessages is a type check task. (#3913)

smillst committed:
Print the underlying exception message. (#3915)

mernst committed:
Change order of parameters so the value being side-effected comes first (#3914)

mernst committed:
Add canary_jobs as an abstraction for Azure Pipelines

mernst committed:
Checkers that issue no warnings can ignore the "allcheckers" prefix (#3920)

mernst committed:
Reorder code for clarity.  No logic changes. (#3921)

mernst committed:
Improve Javadoc style

mernst committed:
`-AwarnUnneededSuppressionsExceptions=REGEX` command-line option

mernst committed:
Change boolean to enum for readability; fixes #3390

mernst committed:
Documentation improvements and a renaming

mernst committed:
Define method ElementUtils.closure

smillst committed:
Fix Javadoc. (#3925)

smillst committed:
Clarify VisitHistory classes. (#3926)

mernst committed:
Update Dockerfiles to Ubuntu 20.10

mernst committed:
Rename ElementUtils.closure to getAllSupertypes

smillst committed:
When building a type tree, don't recur infinitely on wildcards

smillst committed:
Captured types in enhanced for loops

mernst committed:
Move declaration closer to reassignment

mernst committed:
Change order of code.  No logic changes.

mernst committed:
Remove reference to Java 8; tweak text about compilation speed

mernst committed:
Javadoc fixes

mernst committed:
Add documentation and shorten logic of ignoreUninferredTypeArgument() (#3934)

mernst committed:
Documentation improvements (#3937)

mernst committed:
Performe WPI before the store has been side-effected (#3938)

mernst committed:
StubParser minor refactoring  (#3941)

mernst committed:
Rename `AnnotatedTypeMirror.actualType` to `underlyingType` (#3942)

mernst committed:
Remove a redundant method

mernst committed:
Make the visualization of stores deterministic

mernst committed:
Tweak documentation of qualifier polymorphism (#3945)

mernst committed:
Add method getUninitializedFields()

mernst committed:
Shell script style

mernst committed:
Rename `Contractsutils` to `ContractsFromMethod`

smillst committed:
Ignore declaration annotations when issuing nullness.on.primitive warning; fixes #3935

mernst committed:
Define method getOverriddenMethods

mernst committed:
Tweak description of type refinement

mernst committed:
Remove irrelevant information from bin/README (#3950)

mernst committed:
Abstract out methods ignoreFieldInWPI and updateFieldFromType (#3951)

smillst committed:
Release prep.

smillst committed:
Move CalledMethods Checker test to the correct location. (#3954)

smillst committed:
Can't use python3.

smillst committed:
Fix links.

smillst committed:
new release 3.8.0

smillst committed:
Prep for next release.

mernst committed:
Refactor StubParser.getAnnotation to eliminate recursion (#3957)

mernst committed:
Nested class StubAnnotations represents the annotations in a stub file (#3958)

mernst committed:
Better instructions for diffing contributors.tex

mernst committed:
StubParser cleanups

mernst committed:
Rename field `typeName` to `typeBeingParsed`

mernst committed:
Rename StubParser to AnnotationFileParser

mernst committed:
Add test of getInterfaces and purity

mernst committed:
Define WholeProgramInferenceScenes.updateAnnotationSetInScene()

mernst committed:
Eliminate variables that are used 0 or 1 times (#3964)

jwaataja committed:
Make packageAnnos a field in AnnotationFileParser

mernst committed:
Add another example for `@DotSeparatedIdentifier`

mernst committed:
Clarify the meaning of `@SuppressWarnings` on a variable declaration

mernst committed:
Define method createWholeProgramInference

mernst committed:
Define method `internalReprOfReceiver()` (#3973)

kelloggm committed:
allow users to run the Called Methods Checker without the Returns Receiver Checker, if desired (#3972)

mernst committed:
Prevent duplicate annotations in .class files; fixes #3956

mernst committed:
Rename Receiver to JavaExpression

mernst committed:
Correct method name

mernst committed:
Move WPI tests to `checker` package (#3978)

mernst committed:
Tweak recommendation about getKind(), and add links to Javadoc

mernst committed:
Make calls to `new AnnotationBuilder()` more concise (#3979)

artempyanykh committed:
Fix CFG builder loosing catch blocks in try-with-resource

mernst committed:
Document when getAnnotationInHierarchy return null (#3980)

mernst committed:
Improve documentation about equality (#3981)

mernst committed:
Make annotationMirrorToAnnotation throw an exception rather than returning null

mernst committed:
Documentation about WPI tests

jwaataja committed:
Use Java 11 for collecting Java files

mernst committed:
Add contributor

smillst committed:
Use util method to check if type has a receiver. (#3986)

smillst committed:
Copy annotations to the enclosing type of an erased type. (#3985)

mernst committed:
Deprecate "class element" in favor of "type element" (#3989)

mernst committed:
Implement getAnnotationFieldType without depending on values

mernst committed:
Don't create a separate class just for utility methods

mernst committed:
Make pathname explicit in example

mernst committed:
Add "NullPointerException" to improve searchability

mernst committed:
Improve formatting of examples

mernst committed:
Clarify documentation

mernst committed:
annotationMirrorToAnnotation returns a non-null value

mernst committed:
Advice about asking for help with open issues

mernst committed:
Reduce indentation

mernst committed:
Spell-check

mernst committed:
Fix whitespace

mernst committed:
Test case for issue #4007

mernst committed:
Don't issue format.indirect.arguments for wrapped format calls (#3996)

mernst committed:
Add field WholeProgramInferenceScenesStorage.atypeFactory (#3998)

mernst committed:
Add `prepareSceneForWriting` hook

mernst committed:
Rename getParamTypes to getArgTypes (#4001)

mernst committed:
Declare an alias for `@FormatMethod` (#4002)

mernst committed:
Clean up ConversionCategory.toString (#4010)

mernst committed:
Diagnostics when files are missing (#4012)

mernst committed:
Fix checkername prefix in SuppressWarnings strings

smillst committed:
Add annotations that may be missing from inferred types. (#3987)

mernst committed:
Remove unnecessary warning suppressions

mernst committed:
Simplify constructor calls and define `isAssignableFrom`

t-8ch committed:
j.u.Optional: test more methods; fixes #4005

Co-authored-by: Michael Ernst <mernst@cs.washington.edu>

mernst committed:
Remove the canary_jobs abstraction

This removal reduces latency by giving less chance for later jobs to get their first jobs into the queue before this one's last jobs get in the queue.

smillst committed:
Primitives are always "from byte code"

mernst committed:
Don't ignore compiler output in issue1585 test

mernst committed:
Deprecate `addAliasedAnnotation`; use `addAliasedTypeAnnotation`

mernst committed:
Define method TypesUtils.getClassFromType

mernst committed:
Improve AnnotationFileParser diagnostics (#4017)

mernst committed:
Further support for com.google.errorprone.annotations.FormatMethod

mernst committed:
Rename `ThisLiteral` to `This`, because it's not a literal

mernst committed:
Handle calls to format with a null varargs array

mernst committed:
Remove SKIP-REQUIRE-JAVADOC file

mernst committed:
Improve location for error messages

mernst committed:
Adapt to StubParser changes

mernst committed:
Test case of class annotation without package declaration

mernst committed:
Add `@AnnotatedFor` to stub files

smillst committed:
Update the Checker Framework Gradle plugin in the examples. (#4020)

mernst committed:
Improve attributeEquals (#4026)

kelloggm committed:
Merge branch 'master' into fix-issue-2452

kelloggm committed:
fix errors intro'd by mergE
